### ISSUE TITLE
RecursionError when using attr.asdict - intended?

### ISSUE THREAD

**Issue by wimglenn** [python-attrs/attrs#740]
Suppose we use attrs instances to represent graphs:

```
>>> @attr.s
... class Node:
...     val = attr.ib()
...     linked_node = attr.ib(default=None)
... 
>>> n = Node(1)
>>> attr.asdict(n)
{'val': 1, 'linked_node': None}
>>> n.linked_node = n
>>> n
Node(val=1, linked_node=...)
```

Although the repr handles it, `attr.asdict` does not detect and resolve cycles - should it?

```
>>> d = attr.asdict(n)
...
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
...
.venv/lib/python3.9/site-packages/attr/_funcs.py in asdict(inst, recurse, filter, dict_factory, retain_collection_types, value_serializer)
     60         if recurse is True:
     61             if has(v.__class__):
---> 62                 rv[a.name] = asdict(
     63                     v,
     64                     True,

RecursionError: maximum recursion depth exceeded while calling a Python object
```

Maybe this the intended behavior (if so, please close this issue), I'm not sure. But I half-expected a self-referencing dict to be returned:

```
>>> d
{'val': 1, 'linked_node': {...}}
```

**Comment by dam5h** [comment]
I have the same issue.

**Comment by wbolster** [comment]
tbh i don't think this is worth ‘fixing’. in the case of `repr()`, mostly useful for developing/debugging, _not crashing_ is useful behaviour, and consistent with how `repr()` behaves for built-ins like dictionaries:

``` python
>>> d = {"a": 123}
>>> d["a"] = d
>>> d
{'a': {...}}
```

however, throw some `json` serialisation into the mix and it will 💥:

``` python
>>> import json
>>> json.dumps(d)
Traceback (most recent call last):
  Input In [14] in <cell line: 1>
    json.dumps(d)
  File /usr/lib/python3.10/json/__init__.py:231 in dumps
    return _default_encoder.encode(obj)
  File /usr/lib/python3.10/json/encoder.py:199 in encode
    chunks = self.iterencode(o, _one_shot=True)
  File /usr/lib/python3.10/json/encoder.py:257 in iterencode
    return _iterencode(o, 0)
ValueError: Circular reference detected
```

it raises `ValueError` (b/c `json` does its own detection), while `attr.asdict()` raises `RecursionError`, but that's an implementation detail.

the key difference here is that, unlike `repr()`, there is no single obvious solution here, and hence raising is the right way to go.

tempted to ‘wontfix’. thoughts?

**Comment by wimglenn** [comment]
I don't think json encoding is a fair analogy, because it is not possible to represent circular references in a json serialization. It _is_ possible in a Python dict, so why shouldn't `attr.asdict` just do that?

If raising an exception _is_ the right way to go, it should still be a controlled failure mode rather than consuming resources and eventually blowing the call stack out with recursion error - this way, there's the possibility to give some context about where the circular reference is coming from in the error message.

So, I think whether attrs returns a dict here or raises an exception, the cycle detection would still be beneficial either way.

**Comment by wbolster** [comment]
fair enough, producing a recursive dict would be a reasonable behaviour indeed. but would this *always* be desired?

(though in practice dict conversion is often used as a quick way to serialise which may fail anyway)

@Tinche what does `cattrs` do here? (and why?)

**Comment by Tinche** [comment]
I don't think we do anything special apart from raising a RecursionError. Serializing recursive data is tricky and requires a completely different schema approach (you need to switch to some sort of pointer abstraction) than what's normally expected in JSON/bson etc. And this behavior would have to be opt-in so 99% of other, more usual cases don't take a speed hit.

I guess the question is how much more complex and slower would asdict have to become to support this?

**Comment by wimglenn** [comment]
I don’t think it is that tricky in practice, you just keep track of visited ids as you go. The repr is already doing this, as mentioned, so take a look in there for an example.

**Comment by wbolster** [comment]
sure, a pr for testing may help.

the current `repr()` implementation uses `threading.local()`, with a `.already_repring` that holds a set of currently ‘repr'ed’ `id(...)` values, because:

```
# Thread-local global to track attrs instances which are already being repr'd.
# This is needed because there is no other (thread-safe) way to pass info
# about the instances that are already being repr'd through the call stack
# in order to ensure we don't perform infinite recursion.
```

i suspect for `.asdict()` this may not be needed, and a ‘memo’ argument that's passed down would work, similar to how `copy.copy()` works?

**Comment by Darkdragon84** [comment]
The same problem also exists for `__eq__`. dicts with self references are able to compare and return the expected
```
>>> a = dict()
>>> a[1] = a
>>> a == a
True
```
It would be really nice to have a recursion detection for all these recursive methods :-)
